### PR TITLE
WaitHelper threshold for logging

### PR DESCRIPTION
**Pull request by SpComb**
Implement a new `WaitHelper#wait_until(threshold: 1.0)`, which:

* Does not log anything if it does not wait, i.e. the block yields truthy immediately without any sleeps
* Logs debug messages after 0.5, 1.0, 1.0, 2.0, 4.0, ... seconds of waiting
* Logs an info message if the total wait time was over the threshold
* Logs a warning message if the wait timed out

This reduces the high level of DEBUG spam from short-interval waits, and also provides useful INFO-level messages for slow waits.

Copy-pasted to both agent and server codebases.

```
I, [2017-04-05T18:11:52.437912 #31448]  INFO -- GridServiceInstanceDeployer: Deploying service instance development/null/redis-fail-1 to node core-01 at 2017-04-05 15:11:52 UTC...
D, [2017-04-05T18:11:52.982966 #31448] DEBUG -- GridServiceInstanceDeployer: waiting 1.0s of 300.0s until: service development/null/redis-fail-1 is running on node development/core-01 at 2017-04-05 15:11:52 UTC
D, [2017-04-05T18:11:53.484436 #31448] DEBUG -- GridServiceInstanceDeployer: waiting 1.5s of 300.0s until: service development/null/redis-fail-1 is running on node development/core-01 at 2017-04-05 15:11:52 UTC
D, [2017-04-05T18:11:54.487590 #31448] DEBUG -- GridServiceInstanceDeployer: waiting 2.5s of 300.0s until: service development/null/redis-fail-1 is running on node development/core-01 at 2017-04-05 15:11:52 UTC
D, [2017-04-05T18:11:56.494978 #31448] DEBUG -- GridServiceInstanceDeployer: waiting 4.5s of 300.0s until: service development/null/redis-fail-1 is running on node development/core-01 at 2017-04-05 15:11:52 UTC
D, [2017-04-05T18:12:00.514560 #31448] DEBUG -- GridServiceInstanceDeployer: waiting 8.5s of 300.0s until: service development/null/redis-fail-1 is running on node development/core-01 at 2017-04-05 15:11:52 UTC
D, [2017-04-05T18:12:08.548030 #31448] DEBUG -- GridServiceInstanceDeployer: waiting 16.6s of 300.0s until: service development/null/redis-fail-1 is running on node development/core-01 at 2017-04-05 15:11:52 UTC
D, [2017-04-05T18:12:24.752304 #31448] DEBUG -- GridServiceInstanceDeployer: waiting 32.8s of 300.0s until: service development/null/redis-fail-1 is running on node development/core-01 at 2017-04-05 15:11:52 UTC
D, [2017-04-05T18:12:56.884714 #31448] DEBUG -- GridServiceInstanceDeployer: waiting 64.9s of 300.0s until: service development/null/redis-fail-1 is running on node development/core-01 at 2017-04-05 15:11:52 UTC
I, [2017-04-05T18:13:10.439213 #31448]  INFO -- GridServiceInstanceDeployer: waited 78.0s of 300.0s until: service development/null/redis-fail-1 is running on node development/core-01 at 2017-04-05 15:11:52 UTC yielded #<GridServiceInstance:0x007fb4ccf3cf90>
W, [2017-04-05T18:13:10.439639 #31448]  WARN -- GridServiceInstanceDeployer: Failed to deploy service instance development/null/redis-fail-1 to node core-01: GridServiceInstanceDeployer::ServiceError: Celluloid::DeadActorError: attempted to call a dead actor: ensure_image
```

## TODO
* Tweak threshold for different classes/waits?

### CODE REVIEW

**Comment by SpComb** [comment]
If the design looks good, then I'll also update the server's `WaitHelper` with the same threshold.

**Comment by SpComb** [comment]
Perhaps it should also skip the `waited ...` debug logging if the condition is immediately true, without any sleep?

```
D, [2017-04-05T12:38:53.157137 #1] DEBUG -- Kontena::RpcClient: waited 0.0s of 30.0s until: websocket client is connected yielded true
D, [2017-04-05T12:38:53.167724 #1] DEBUG -- Kontena::RpcClient: waited 0.0s of 30.0s until: request /node_service_pods/set_state has response wth id=757451937 yielded [{}, nil]
```

The first wait is kinda pointless logging.

**Comment by SpComb** [comment]
PTAL @jakolehm @jnummelin

File: server/spec/helpers/wait_helper_spec.rb [hunk 1]
```diff
@@ -18,24 +18,34 @@
...
 
       value = subject.wait_until { true }
 
       expect(value).to be_truthy
     end
 
-    it 'sleeps between retries and logs debug' do
-      expect(subject).to receive(:debug).with('waiting 0.5s of 2.0s until: something that is true the second time')
-      expect(subject).to receive(:debug).with('waited 0.5s until: something that is true the second time yielded true')
+    it 'sleeps between retries and logs debug while under threshold' do
```
Comment (jakolehm) [inline server/spec/helpers/wait_helper_spec.rb]: We probably should use `timecop` (or similar) gem here.. so that we can timetravel without slowing specs.

Comment (SpComb) [inline server/spec/helpers/wait_helper_spec.rb]: It already mocks out the `sleep` and keeps an mock clock, so these specs shouldn't be slowed.

Comment (jakolehm) [inline server/spec/helpers/wait_helper_spec.rb]: Missed that :)

**Pull request merged**